src/lib/metrics.js

```diff
@@ -0,0 +1,334 @@
+      cm3,
+      cd1,
+      cd2
+    });
+  },
+
+  themeChangeBackground(backgroundId) {
+    cm2++;
+    this.setThemeChanged(true);
+    this.sendEvent({
+      ec: 'theme-change',
+      ea: 'select-background',
+      el: backgroundId,
+      cm1,
+      cm2,
+      cm3,
+      cd1,
+      cd2,
+      cd3,
+      cd11
+    });
+  },
+
+  themeChangeColor(colorName) {
+    cm3++;
+    this.setThemeChanged(true);
+    this.sendEvent({
+      ec: 'theme-change',
+      ea: 'select-color',
+      el: colorName,
+      cm1,
+      cm2,
+      cm3,
+      cd1,
+      cd2,
+      cd3
+    });
+  },
+
+  shareClick() {
+    this.sendEvent({
+      ec: 'share-engagement',
+      ea: 'button-click',
+      cm1,
+      cm2,
+      cm3,
+      cd1,
+      cd2,
+      cd3,
+      cd4,
+      cd5,
+      cd6,
+      cd7,
+      cd8,
+      cd9,
+      cd10,
+      cd11
+    });
+  },
+
+  linkClick(el) {
+    // if el === download-firefox, add the following dimensions to this event
+    const downloadFirefoxDimensions =
+      el !== 'download-firefox' ? {} : { cd5, cd6, cd7, cd8, cd9, cd10, cd11 };
+    this.sendEvent({
+      ec: 'link-engagement',
+      ea: 'click',
+      el,
+      cm1,
+      cm2,
+      cm3,
+      cd1,
+      cd2,
+      cd3,
+      cd4,
+      ...downloadFirefoxDimensions
+    });
+  },
+
+  receiveTheme(action, theme) {
+    this.sendEvent({
+      ec: 'receive-theme',
+      ea: 'button-click',
+      el: action,
+      cd1,
+      ...this.themeToDimensions(theme)
+    });
+  },
+
+  finishVisit() {
+    this.sendEvent({
+      ec: 'finish-visit',
+      ea: 'leave',
+      cm2,
+      cm3,
+      cd1,
+      cd2,
+      cd3,
+      cd4,
+      cd5,
+      cd6,
+      cd7,
+      cd8,
+      cd9,
+      cd10,
+      cd11
+    });
+  },
+
+  sendEvent(params) {
+    if (!GA_TRACKING_ID || !clientUUID || navigator.doNotTrack === 1) {
+      return;
+    }
+
+    const event = {
+      v: 1,
+      aip: 1, // anonymize user IP addresses
+      an: GA_APP_ID,
+      av: GA_APP_VERSION,
+      tid: GA_TRACKING_ID,
+      cid: clientUUID,
+      t: 'event',
+      ...params
+    };
+
+    log('sendEvent', clientUUID, event);
+
+    // Form-encode the event
+    const encoded = Object.entries(event)
+      .map(
+        ([key, value]) =>
+          `${encodeURIComponent(key)}=${encodeURIComponent(value)}`
+      )
+      .join('&');
+
+    // Send using background beacon, if available. Otherwise use fetch()
+    if (typeof navigator !== 'undefined' && 'sendBeacon' in navigator) {
+      navigator.sendBeacon(GA_URL, encoded);
+    } else if (typeof fetch !== 'undefined') {
+      fetch(GA_URL, {
+        method: 'POST',
+        mode: 'cors',
+        headers: {
+          'Content-Type': 'application/x-www-form-urlencoded; charset=utf-8'
+        },
+        body: encoded
+      })
+        .then(() => log(`Sent GA message via fetch: ${encoded}`))
+        .catch(err => log(`GA sending via fetch failed: ${err}`));
+    }
```

Comment:
Do we need a fallback if [`fetch()` is not supported](https://developer.mozilla.org/docs/Web/API/Fetch_API#Browser_compatibility)?

Reply:
Yeah, a polyfill might be a good idea